### PR TITLE
[RF] Modernise RooFit's Category Classes

### PR DESCRIPTION
**Pull request by hageboeck**
**For reviewers**:
The most important commit is the first, and specifically the files
`RooAbsCategory.h, RooCategory.h`
The former is the base class of all categories, the latter the user-facing implementation.

Here's what happened:
RooFit categories were using the class RooCatType (: public TObject)
instead of an integer to represent category states.
This class contained an integer and a char[256] to store a category index
and a (possibly truncated) state name. This prevented fast batch access,
and increased the overhead for data storage and data loading.

From now on, categories are represented as integers.
Every state can be denoted by index or name, but instead of storing
*both packed into an object* in each row of the dataset, only the index is stored, and the
mapping from index to state name is stored in the category instance.

Since all the category classes were happily using the RooCatType, several
parts had to be touched to convert everything to using integers. This feature
has been cooking since March, and the interface for category classes was
discussed in the PPP, but nevertheless, have a double look at the above-mentioned
headers.

In more detail:
- Replace RooCatType by an integer.
- Replace names saved in each row of dataset with RooCatType by a map in
  RooAbsCategory that maps state names to numbers. This saves 64 bytes
  + the size of TObject for each event, and makes loading strings in each
  event unnecessary.
- Provide legacy iterator for iterating through states instances of
  RooCatType, which will be created on the fly.
- Make RooVectorDataStore store integers instead of RooCatType instances.
  - With this, also reduce the amount of unnecessary members in RooVectorDataStore.
  - Outline functions that polluted the header of RooVectorDataStore.

Updates for category interfaces:
- Provide new interfaces for categories to work without RooCatType.
- Add bracket operator and constructor arguments to define multiple
  category states at once or using category["stateName"] = 1.
- Mark legacy interfaces exposing RooCatType as deprecated in doxygen.
- Stop including RooCatType when R__LESS_INCLUDES is set.
- Flag legacy interfaces with R__SUGGEST_ALTERNATIVE.
- Don't create RooCatType instances, unless users continue to use
  the legacy interface.
- Provide functions hasIndex() and hasLabel() to check validity of
  index/label.
- Provide function states() with direct access to map of category names
  to category numbers.
- Add functions getCurrent{Index|Label} to make clear that they access
  class state.

- Prevent clearing of shapeDirty in RooAbsCategory::evaluate(), since
  derived category classes might need to recompute their shapes if one
  of their input categories change.

Changes in special category classes:
- Consolidate RooMultiCategory and RooSuperCategory. These share almost
  all the code, so RooSuperCategory will now use a RooMultiCategory for
  state definition and retrieval. The only additional functionality for
  RooSuperCategory is that it can assign states to its subcategories.
- Clean up header of RooMappedCategory. Outline as much as possible.

### CODE REVIEW

**Comment by eguiraud** [comment]
Hi, I'll take a look today, probably in the afternoon. @hageboeck looks like there are conflicts

**Comment by hageboeck** [comment]
Due to Oksana's update of gtest. Rebasing gave me the opportunity to add a fix and squash commits. Pushing soon.

File: roofit/roofitcore/inc/LinkDef.h [hunk 1]
```diff
@@ -29,12 +29,21 @@
...
+           _insertionOrder.push_back(catType->getVal()); \
+         }}";
+#pragma read sourceClass="RooAbsCategory" targetClass="RooAbsCategory" version="[2]" \
+  include="RooFitLegacy/RooCatTypeLegacy.h" \
+  source="std::vector<RooCatType*> _types" target="_stateNames,_insertionOrder" \
+  code="{for (const auto catType : onfile._types) { _stateNames[catType->GetName()] = catType->getVal();\
+                                                    _insertionOrder.push_back(catType->getVal());\
+                                                  } }";
+#pragma read sourceClass="RooAbsCategory" targetClass="RooAbsCategory" version="[1-2]" include="RooFitLegacy/RooCatTypeLegacy.h" \
+  source="RooCatType _value" target="_currentIndex" code="{ _currentIndex = onfile._value.getVal(); }"
```
Comment (eguiraud) [inline roofit/roofitcore/inc/LinkDef.h:46]: just curious..what on earth does this do?

File: roofit/roofitcore/inc/RooAbsCategory.h [hunk 2]
```diff
@@ -86,38 +97,101 @@ class RooAbsCategory : public RooAbsArg {
...
+    return stateNames().size();
   }
 
-  std::vector<RooCatType*>::const_iterator end() const {
-    return _types.cend();
+  /// Access the map of state names to index numbers. Triggers a recomputation
+  /// if the shape is dirty.
+  const std::map<std::string, value_type>& stateNames() const {
+    return const_cast<RooAbsCategory*>(this)->stateNames();
+  }
```
Comment (eguiraud) [inline roofit/roofitcore/inc/RooAbsCategory.h]: this `const_cast` makes the `const` method thread-unsafe. if it's possible that someone might, in the future, care about thread-safety of `RooAbsCategory`, consider marking whatever state `clearShapeDirty` modifies `mutable` and protect accesses to it with a lock

File: roofit/roofitcore/inc/RooAbsCategoryLValue.h [hunk 3]
```diff
@@ -30,15 +30,24 @@ class RooAbsCategoryLValue : public RooAbsCategory, public RooAbsLValue {
...
-  virtual Bool_t setIndex(Int_t index, Bool_t printError=kTRUE) = 0 ;
-  virtual Bool_t setLabel(const char* label, Bool_t printError=kTRUE) = 0 ;
+  ////////////////////////////////////////////////////////////////////////////////
+  /// Set value by specifying the index code of the desired state.
+  /// If printError is set, a message will be printed if
+  /// the specified index does not represent a valid state.
+  /// \return bool to signal an error.
+  virtual bool setIndex(value_type index, bool printError = true) = 0;
+  bool setOrdinal(unsigned int index);
+  virtual bool setLabel(const char* label, Bool_t printError=kTRUE) = 0 ;
```
Comment (eguiraud) [inline roofit/roofitcore/inc/RooAbsCategoryLValue.h:40]: trailing space after 0 :smile:

File: roofit/roofitcore/inc/RooAbsCategoryLValue.h [hunk 4]
```diff
@@ -65,17 +74,22 @@ class RooAbsCategoryLValue : public RooAbsCategory, public RooAbsLValue {
     return kTRUE; 
   }
 
-  // I/O streaming interface (machine readable)
-  virtual Bool_t readFromStream(std::istream& is, Bool_t compact, Bool_t verbose=kFALSE) ;
-  virtual void writeToStream(std::ostream& os, Bool_t compact) const ;
+  /// Not implemented.
+  virtual Bool_t readFromStream(std::istream&, Bool_t, Bool_t) { return true; }
+  /// Not implemented.
+  virtual void writeToStream(std::ostream&, Bool_t) const { }
```
Comment (eguiraud) [inline roofit/roofitcore/inc/RooAbsCategoryLValue.h:80]: I realize it was already like this, but this pattern is evil...you can _actually_ leave the methods unimplemented if pure virtual is not an option

File: roofit/roofitcore/inc/LinkDef.h (hunk 1 above)
Comment (hageboeck) [reply roofit/roofitcore/inc/LinkDef.h:46]: It's schema evolution. These are read rules telling ROOT how to read an old class version on file that stores category states in the 284-bytes-too-large `RooCatType`. You basically say:
Get me the RooCatType from file, and I do with it what's necessary.

File: roofit/roofitcore/inc/RooAbsCategory.h (hunk 2 above)
Comment (hageboeck) [reply roofit/roofitcore/inc/RooAbsCategory.h]: That's true. Thread safety is not an issue yet, but surely, I shouldn't make things worse ...

File: roofit/roofitcore/inc/RooAbsCategoryLValue.h (hunk 4 above)
Comment (hageboeck) [reply roofit/roofitcore/inc/RooAbsCategoryLValue.h:80]: 👍 I could actually remove these!

**Pull request merged**